crates/subspace-service/src/sync_from_dsn/import_blocks.rs

```diff
@@ -239,69 +245,56 @@ async fn download_and_reconstruct_blocks<PV>(
     segment_index: SegmentIndex,
     piece_provider: &PieceProvider<PV>,
     reconstructor: &mut Reconstructor,
+    dsn_sync_parallelism_level: usize,
 ) -> Result<Vec<(BlockNumber, Vec<u8>)>, sc_service::Error>
 where
     PV: PieceValidator,
 {
-    debug!(%segment_index, "Retrieving pieces of the segment");
+    debug!(%segment_index, %dsn_sync_parallelism_level, "Retrieving pieces of the segment");
 
-    let semaphore = &Semaphore::new(RecordedHistorySegment::NUM_RAW_RECORDS);
+    let semaphore = &Semaphore::new(dsn_sync_parallelism_level);
 
     let mut received_segment_pieces = segment_index
         .segment_piece_indexes_source_first()
         .into_iter()
-        .map(|piece_index| {
-            // Source pieces will acquire permit here right away
-            let maybe_permit = semaphore.try_acquire().ok();
-
-            async move {
-                let permit = match maybe_permit {
-                    Some(permit) => permit,
-                    None => {
-                        // Other pieces will acquire permit here instead
-                        match semaphore.acquire().await {
-                            Ok(permit) => permit,
-                            Err(error) => {
-                                warn!(
-                                    %piece_index,
-                                    %error,
-                                    "Semaphore was closed, interrupting piece retrieval"
-                                );
-                                return None;
-                            }
-                        }
-                    }
-                };
-                let maybe_piece = match piece_provider
-                    .get_piece(
-                        piece_index,
-                        RetryPolicy::Limited(PIECE_GETTER_RETRY_NUMBER.get()),
-                    )
-                    .await
-                {
-                    Ok(maybe_piece) => maybe_piece,
-                    Err(error) => {
-                        trace!(
-                            %error,
-                            ?piece_index,
-                            "Piece request failed",
-                        );
-                        return None;
-                    }
-                };
+        .map(|piece_index| async move {
+            let _permit = match semaphore.acquire().await {
```

Comment:
Great, now we always retrieve 256 pieces even if we need just 128 and we retrieve random pieces rather than source first, which impacts reconstruction performance.